### ISSUE TITLE
v2.1 SEQAN_BUILD_SYSTEM=SEQAN_RELEASE_LIBRARY make test fails

### ISSUE THREAD

**Issue by mr-c** [seqan/seqan#1478]
```
mcrusoe@mrcdev:~/src/seqan-build$ cmake ../seqan-seqan-v2.1.0/ -DSEQAN_BUILD_SYSTEM=SEQAN_RELEASE_LIBRARY
-- The CXX compiler identification is GNU 5.3.1
-- Check for working CXX compiler: /usr/lib/ccache/c++
-- Check for working CXX compiler: /usr/lib/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Initializing SeqAn Build System...
-- Performing Test CXX11_DETECTED
-- Performing Test CXX11_DETECTED - Success
-- Performing Test CXX14_DETECTED
-- Performing Test CXX14_DETECTED - Success
--   Selected repository dir: /home/mcrusoe/src/seqan-seqan-v2.1.0
-- No revision system found.
--   Repository date not determined.
--   Repository revision not determined.
-- Configuring apps
-- Configuring apps
-- Configuring dox
-- Found PythonInterp: /usr/bin/python (found version "2.7.11") 
-- Configuring manual
-- Found Sphinx: /usr/bin/sphinx-build  
-- Configuring util/py_lib
--   adding nosetests for seqan.dox
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mcrusoe/src/seqan-build
```
```
mcrusoe@mrcdev:~/src/seqan-build$ make test
Running tests...
Test project /home/mcrusoe/src/seqan-build
    Start 1: build_dox
1/3 Test #1: build_dox ........................   Passed   25.00 sec
    Start 2: build_manual
2/3 Test #2: build_manual .....................***Failed    0.37 sec
    Start 3: test_py_lib_dox
3/3 Test #3: test_py_lib_dox ..................   Passed    0.59 sec

67% tests passed, 1 tests failed out of 3

Total Test time (real) =  25.97 sec

The following tests FAILED:
          2 - build_manual (Failed)
Errors while running CTest
Makefile:61: recipe for target 'test' failed
make: *** [test] Error 8
```
```
mcrusoe@mrcdev:~/src/seqan-build$ uname -a
Linux mrcdev 4.3.0-1-amd64 #1 SMP Debian 4.3.3-7 (2016-01-19) x86_64 GNU/Linux
mcrusoe@mrcdev:~/src/seqan-build$ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux unstable (sid)
Release:        unstable
Codename:       sid
```

**Comment by mr-c** [comment]
```
2/3 Testing: build_manual
2/3 Test: build_manual
Command: "/usr/bin/sphinx-build" "-W" "-n" "-b" "html" "-d" "/home/mcrusoe/src/seqan-build/manual/doctrees" "/home/mcrusoe/src/seqan-seqan-v2.1.0/manual/source" "/home/mcrusoe/src/seqan-build/manual/html"
Directory: /home/mcrusoe/src/seqan-build/manual
"build_manual" start time: Feb 17 12:33 PST
Output:
----------------------------------------------------------
Running Sphinx v1.3.5

Extension error:
Could not import extension sphinxcontrib.bibtex (exception: No module named sphinxcontrib.bibtex)
<end of output>
Test time =   0.36 sec
----------------------------------------------------------
Test Failed.
"build_manual" end time: Feb 17 12:33 PST
"build_manual" time elapsed: 00:00:00
----------------------------------------------------------
```

**Comment by mr-c** [comment]
For that matter, `make test` fails with no `SEQAN_BUILD_SYSTEM` defined due to missing test binaries.

**Comment by marehr** [comment]
Thank you for your bug report. I will look into it.

**Comment by marehr** [comment]
Could you run ```ctest --output-on-failure``` instead of ```make test```? It would be interesting why the test fails.

**Comment by mr-c** [comment]
@marehr output of the test failure is in https://github.com/seqan/seqan/issues/1478#issuecomment-185392826 the Python module sphinxcontrib.bibtex doesn't seem to be bundled with the source.

**Comment by h-2** [comment]
@mr-c We will look into the failing tests, I have to admit we never run `make test`, but ctest directly. Also IMO the manual shouldn't be part of `SEQAN_RELEASE_LIBRARY` at all.

In any case, if you want to ship the seqan library, I recommend you use the library packages from
http://packages.seqan.de/
as they are much smaller and require no build dependencies at all (because there is no build). They come with prebuilt API documentation (but not the manual).

**Comment by mr-c** [comment]
Thanks for the response!

We'd prefer to build and run the tests on all of Debian architectures while
making the packages to ensure that they will work.

For example, here is the output of the build for SeqAn 1.4.1:
https://buildd.debian.org/status/package.php?p=seqan

Since we are also shipping the apps (though in a seperate package), I've
switched to building the entire release for the Debian packages.

Joi, 18 feb. 2016, 11:13, Hannes Hauswedell <notifications@github.com> a
scris:

> @mr-c <https://github.com/mr-c> We will look into the failing tests, I
> have to admit we never run make test, but ctest directly. Also IMO the
> manual shouldn't be part of SEQAN_RELEASE_LIBRARY at all.
>
> In any case, if you want to ship the seqan library, I recommend you use
> the library packages from
> http://packages.seqan.de/
> as they are much smaller and require no build dependencies at all (because
> there is no build). They come with prebuilt API documentation (but not the
> manual).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/seqan/seqan/issues/1478#issuecomment-185638406>.
>
-- 
Michael R. Crusoe     CWL Community Engineer     crusoe@ucdavis.edu
<mcrusoe@msu.edu>
Common Workflow Language project    University of California, Davis
https://impactstory.org/MichaelRCrusoe http://twitter.com/biocrusoe

**Comment by h-2** [comment]
@marehr can you change the top-level CMakeLists.txt to not `add_subdirectory (manual)` for `SEQAN_RELEASE_LIBRARY and SEQAN_RELEASE_APPS, and also change it to not `add_subdirectory (apps)` for SEQAN_RELEASE_LIBRARY [the latter just removes the `-- Configuring apps` line].

And we should make the "make test" target depend on the regular make, so that it automatically builds the things it needs for the test.

@mr-c I still recommend the dedicated library tarballs for library installs, though. Since the library is source only, it doesn't need to ship tests and should not require cmake or python for build or installation. Also the library is strictly BSD-licensed, while most of the Applications are GPL. I will create a "Downstream Tutorial" to explain in which way we recommend packaging.

**Comment by mr-c** [comment]
Thanks @h-2 that worked

Here's the changes I made according to your instructions:
```
mcrusoe@mrcdev:~/src/seqan-seqan-v2.1.0$ git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0334f74..0b86c2f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -132,12 +132,17 @@ if (("${SEQAN_BUILD_SYSTEM}" STREQUAL "SEQAN_RELEASE") OR
     add_subdirectory (demos)
 endif ()
 
-message (STATUS "Configuring apps")
-add_subdirectory (apps)
+if (NOT "${SEQAN_BUILD_SYSTEM}" STREQUAL "SEQAN_RELEASE_LIBRARY")
+    message (STATUS "Configuring apps")
+    add_subdirectory (apps)
+endif ()
 message (STATUS "Configuring dox")
 add_subdirectory (dox)
-message (STATUS "Configuring manual")
-add_subdirectory (manual)
+if ((NOT "${SEQAN_BUILD_SYSTEM}" STREQUAL "SEQAN_RELEASE_LIBRARY") AND
+    (NOT "${SEQAN_BUILD_SYSTEM}" STREQUAL "SEQAN_RELEASE_APPS"))
+    message (STATUS "Configuring manual")
+    add_subdirectory (manual)
+endif ()
 message (STATUS "Configuring util/py_lib")
 add_subdirectory (util/py_lib)
```

And the output
```mcrusoe@mrcdev:~/src/seqan-build$ cmake ../seqan-seqan-v2.1.0/ -DSEQAN_BUILD_SYSTEM=SEQAN_RELEASE_LIBRARY
-- The CXX compiler identification is GNU 5.3.1
-- Check for working CXX compiler: /usr/lib/ccache/c++
-- Check for working CXX compiler: /usr/lib/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Initializing SeqAn Build System...
-- Performing Test CXX11_DETECTED
-- Performing Test CXX11_DETECTED - Success
-- Performing Test CXX14_DETECTED
-- Performing Test CXX14_DETECTED - Success
--   Selected repository dir: /home/mcrusoe/src/seqan-seqan-v2.1.0
--   Determined repository date is 2016-02-18_03:23:21_-0800
--   Determined repository revision is 8cb0e32
-- Configuring dox
-- Found PythonInterp: /usr/bin/python (found version "2.7.11") 
-- Configuring util/py_lib
--   adding nosetests for seqan.dox
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mcrusoe/src/seqan-build
mcrusoe@mrcdev:~/src/seqan-build$ make test
Running tests...
Test project /home/mcrusoe/src/seqan-build
    Start 1: build_dox
1/2 Test #1: build_dox ........................   Passed   28.83 sec
    Start 2: test_py_lib_dox
2/2 Test #2: test_py_lib_dox ..................   Passed    0.67 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =  29.50 sec
```

As for the package building, we (Debian/Debian-med) prefer to build from sources as close to upstream (SeqAn in this case) as possible. If the apps are only released concurrently with the library/tests/demos then it makes the most sense to have a single source Debian package that produces multiple installable (or "binary" in Debian-speak) packages.

If the apps are released separately (out of sync with the library), then yes we would package them separately; though that would be a lot of work to set up.

Regardless of the method chosen, installing the seqan library package (a "binary" package in our language, though it will be pure source code) will not require the presence of cmake or python.

With respect to licensing, we often package source with mixed licensing and we carefully document that using a machine and human readable format. For example, see [the copyright file for seqan 1.4.1](http://sources.debian.net/src/seqan/1.4.1%2Bdfsg-2/debian/copyright/)

**Comment by h-2** [comment]
@mr-c well thats of course up to you how you want to package!

Another thing: to resolve issues with our app packages we build those with static binaries now. Unfortunately I did not introduce a a flag to prevent this for downstream packaging systems (where static binaries are not necessary and usually discouraged). So you might want to comment out
https://github.com/seqan/seqan/blob/master/util/cmake/SeqAnBuildSystem.cmake#L207
(I assume you build the packages with `SEQAN_RELEASE_APPS` or with single `APP:$APPNAME` )

**Comment by mr-c** [comment]
:-) Thanks for the pointer! Yes, a command line option or special target for downstream packagers would be appreciated in a future release.

To confirm: seqan apps are ONLY released in conjunction with seqan library updates; is that correct?

**Comment by h-2** [comment]
> To confirm: seqan apps are ONLY released in conjunction with seqan library updates; is that correct?

Yes, for those apps that are still maintained in the main SeqAn repository. For future versions we want to have entirely separate repositories and we have already moved some apps out / created new apps outside, see 
https://github.com/seqan
Most notably lambda and flexbar. It would of course be great to have these packaged for Debian, as well! But I know it will be a little more work, since they have their own release schedules.

**Comment by mr-c** [comment]
Great to know the roadmap. My main concern is getting seqan-2.1 and updated versions of the existing apps in the archive. I completely agree that having separate repositories for the apps gives you all a lot more flexibility.

So lambda, flexbar, and any other new app won't be packaged until later (at least by me :smile: )

**Comment by xenigmax** [comment]
@marehr 
May I close this by #1560  ?

**Comment by marehr** [comment]
This was fixed by #1560
